homeassistant/components/modbus/switch.py

```diff
@@ -86,15 +93,44 @@ async def async_setup_platform(
 ):
     """Read configuration and create Modbus switches."""
     switches = []
-    if CONF_COILS in config:
-        for coil in config[CONF_COILS]:
-            hub: ModbusHub = hass.data[MODBUS_DOMAIN][coil[CONF_HUB]]
-            switches.append(ModbusCoilSwitch(hub, coil))
-    if CONF_REGISTERS in config:
-        for register in config[CONF_REGISTERS]:
-            hub: ModbusHub = hass.data[MODBUS_DOMAIN][register[CONF_HUB]]
-            switches.append(ModbusRegisterSwitch(hub, register))
 
+    #  check for old config:
+    if discovery_info is None:
+        _LOGGER.warning(
+            "Switch configuration depreciated, will be removed in a future release"
+        )
+        discovery_info = {
+            CONF_NAME: "noName",
+            CONF_SWITCHES: [],
+        }
+        if CONF_COILS in config:
+            discovery_info[CONF_SWITCHES].extend(config[CONF_COILS])
+        if CONF_REGISTERS in config:
+            discovery_info[CONF_SWITCHES].extend(config[CONF_REGISTERS])
+        for entry in discovery_info[CONF_SWITCHES]:
+            if CALL_TYPE_COIL in entry:
+                entry[CONF_ADDRESS] = entry[CALL_TYPE_COIL]
+                entry[CONF_INPUT_TYPE] = CALL_TYPE_COIL
+                del entry[CALL_TYPE_COIL]
+            if CONF_REGISTER in entry:
+                entry[CONF_ADDRESS] = entry[CONF_REGISTER]
+                del entry[CONF_REGISTER]
+                if CONF_REGISTER_TYPE in entry:
+                    entry[CONF_INPUT_TYPE] = entry[CONF_REGISTER_TYPE]
+                    del entry[CONF_REGISTER_TYPE]
+            if CONF_SCAN_INTERVAL not in entry:
+                entry[CONF_SCAN_INTERVAL] = DEFAULT_SCAN_INTERVAL
+        config = None
+
+    for entry in discovery_info[CONF_SWITCHES]:
+        if CONF_HUB in entry:
+            # from old config!
+            discovery_info[CONF_NAME] = entry[CONF_HUB]
+        hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
+        if entry[CONF_INPUT_TYPE] == CALL_TYPE_COIL:
+            switches.append(ModbusCoilSwitch(hub, entry))
+        else:
+            switches.append(ModbusRegisterSwitch(hub, entry))
     async_add_entities(switches)
 
 
```

Comment:
We should inherit from `SwitchEntity` not `ToggleEntity` on the line below.

Reply:
done